designs/2024-baseline-support/README.md

```diff
@@ -0,0 +1,379 @@
+-->
+
+We are storing all the violations that we would like to suppress into a separate file. This file is a JSON file containing the number of errors that must be ignored for each rule in each file. By design, no violations are suppressed - in other words, this feature doesn't affect existing or new projects, unless the developers explicitly suppress one or more violations.
+
+### File format
+
+The JSON file includes details about the file where the violations are found, the rule name and the number of violations. As an example, the following indicates that the file `"src/app/components/foobar/foobar.component.ts"` has one violation for the rule `@typescript-eslint/no-explicit-any` that we want to suppress. All paths are relative to CWD, for portability reasons.
+
+```
+{
+  "src/app/components/foobar/foobar.component.ts": {
+    "@typescript-eslint/no-explicit-any": {
+        count: 1
+    }
+  }
+}
+```
+
+The file is stored in `.eslint-suppressions.json` , unless otherwise specified.
+
+### Suppressing all violations
+
+A new option `--suppress-all` wil be introduced to ESLint CLI. When provided, the JSON file is generated and saved in `.eslint-suppressions.json`. If the file already exists, it gets over-written. Note that this is a boolean flag option (no values are accepted).
+
+``` bash
+eslint --suppress-all ./src
+```
+
+### Suppressing violations of a specific rule
+
+A new option `--suppress-rule [RULE1]` will be introduced to ESLint CLI. When provided, the existing suppressions file will be updated to include any existing violation of the provided rule. The suppressions file will be created if not already exists. Note that this is option can accept an array of string values.
+
+``` bash
+eslint --suppress-rule '@typescript-eslint/no-explicit-any' --suppress-rul '@typescript-eslint/member-ordering' ./src
+```
+
+### Changing the location of the suppressions file
+
+A new option `--suppressions-location` will be introduced to ESLint CLI. When provided, the suppressions file will be loaded and saved to the provided location. Note that this is a string flag option (value is required).
+
+``` bash
+eslint --suppress-all --suppressions-location /home/user/project/mycache ./src
+```
+
+### Maintaining a lean suppressions file
+
+When working with suppressed violations, it's possible to address a violation without updating the suppressions file. This oversight can allow new violations to go unnoticed. To prevent this, eslint can exit with an error code if there are outdated (unmatched) suppressions.
+
+Consider the following scenario:
+
+* The developer runs `eslint --supress-all ./src` to create the suppressions file.
+* Running `eslint ./src` reports no violations and exits with status 0.
+* After fixing a violation, the suppressions file still contains the now-resolved violation.
+* Running `eslint ./src` again reports no violations but exits with a non-zero status code, indicating the suppressions file needs updating.
```

Comment:
In this case, will there be an error message, and what would it look like? Technically, will it be a lint message passed to the formatter along with other linter messages for the file, or a separate output?

Reply:
In `elm-review`, if counts go down, then this file is automatically updated when running the process. The reason I chose to do so is because I don't want to annoy the users when they do the right thing.

Say I fix an issue, I run `eslint` (which on a large project could take a significant amount of time), then I get to hear that I did a good thing by solving problems, but that I need to do the process again. I believe this can be frustrating and I wanted to avoid that.

So we update the file automatically. The downside is that the file might not get checked in properly. To help with that, we recommend that people in CI (or in their test suite) run `elm-review suppress --check-after-tests` which only invokes `git status --short -- review/suppressed/` and reports an error if suppression files should have been committed. (Surprisingly, so far no-one has asked for supporting other VCSs).

Also, we try to display a friendly message when you fixed an issue: "There are still 316 suppressed errors to address, and you just fixed 1!" (I could have added emojis but I also didn't want to go over the top).